tests/parser_tests.py

```diff
@@ -527,6 +530,38 @@ def test_parse_with_extra_words_at_start_and_end_valid(self):
             datetime(2016, 5, 16, 4, 5, 6, 789120),
         )
 
+    # regression test for issue #701
+    # tests cases of a partial match surrounded by punctuation
+    # for the list of valid punctuation, see documentation
+    def test_parse_with_punctuation_fences(self):
+        self.assertEqual(
+            self.parser.parse(
+                "Meet me at my house on Halloween (2019-31-10)", "YYYY-DD-MM"
+            ),
+            datetime(2019, 10, 31),
+        )
+
+        self.assertEqual(
+            self.parser.parse(
+                "Monday, 9. September 2019, 16:15-20:00", "dddd, D. MMMM YYYY"
+            ),
+            datetime(2019, 9, 9),
+        )
+
+        self.assertEqual(
+            self.parser.parse("A date is 11.11.2011.", "DD.MM.YYYY"),
+            datetime(2011, 11, 11),
+        )
+
+        with self.assertRaises(ParserMatchError):
+            self.parser.parse("11.11.2011.1 is not a valid date.", "DD.MM.YYYY")
+
+        with self.assertRaises(ParserMatchError):
+            self.parser.parse(
+                "This date has too many punctuation marks following it (11.11.2011).",
```

Comment:
Looks good overall, thanks for implementing the feedback.

My only concern is that we are making the parsing constraints too strict here by forcing a single piece of punctuation. Should we maybe allow the case of any number of punctuation marks?

Thoughts @systemcatch?

Reply:
I don't think any number would work in my implementation but I can increase it to any finite number you want (say 2 or 3 punctuation marks). I think one is fine; more than two is probably overkill.

Reply:
I tend to agree, but the only thing that concerns me is that this worked pre 0.15.0 (I chose 0.13.0 for example):
```shell
venv ❯ python3
Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 16:52:21)
[Clang 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import arrow
>>> arrow.__version__
'0.13.0'
>>> arrow.get("This date has too many punctuation marks following it 11.11.2011", "DD.MM.YYYY")
<Arrow [2011-11-11T00:00:00+00:00]>
>>> arrow.get("This date has too many punctuation marks following it (11.11.2011)", "DD.MM.YYYY")
<Arrow [2011-11-11T00:00:00+00:00]>
>>> arrow.get("This date has too many punctuation marks following it (11.11.2011).", "DD.MM.YYYY")
<Arrow [2011-11-11T00:00:00+00:00]>
```

This is definitely an improvement, but for full pre-0.15.0 behavior while still containing improvements, we probably need to add support for any number of punctuation marks. Curious, why would finite numbers work but not infinite (e.g. with the `+` quantifier in regex)?

Reply:
I'm not sure if you saw my edit, but I can implement it easily following the date, but preceding doesn't work because [you can't use quantifiers with lookbacks](https://stackoverflow.com/questions/9030305/regular-expression-lookbehind-doesnt-work-with-quantifiers-or).

Reply:
Oh that is interesting, apologies I missed that. Any idea how we could possibly do this without look behinds?

We can merge this in for sure since it is definitely a good fix, but I am wondering if we could potentially do a follow-up to this using another regex technique that allows for more flexibility.

Reply:
I just learned regex to fix this issue, so I'm still a beginner. From my limited research while creating the fix (I was trying to do what you're suggesting at first too), I couldn't find a solution. It's probably still worth making a new issue to see if someone with more regex knowledge can solve it.

A half-solution is just allowing a finite but absurdly high number of punctuation marks.

Reply:
@jadchaar  I think we merge this in and I'll open a new issue to discuss punctuation handling.

Reply:
Agreed. Merging!